google-cloud-firestore/src/main/java/com/google/cloud/firestore/TransactionOptions.java

```diff
@@ -83,10 +116,205 @@ public static TransactionOptions create(@Nonnull Executor executor) {
+  }
+
+  @Nonnull
+  public static ReadOnlyOptionsBuilder readOnlyOptionsBuilder() {
+    return Builder.readOnlyBuilder();
+  }
+
+  public abstract static class Builder<T, B extends Builder<T, B>> {
+    @Nullable protected Executor executor;
+
+    protected Builder(@Nullable Executor executor) {
+      this.executor = executor;
+    }
+
+    @Nullable
+    public Executor getExecutor() {
+      return executor;
+    }
+
+    @Nonnull
+    @SuppressWarnings("unchecked")
+    public B setExecutor(@Nullable Executor executor) {
+      this.executor = executor;
+      return (B) this;
+    }
+
+    @Nonnull
+    public abstract TransactionOptions build();
+
+    static ReadOnlyOptionsBuilder readOnlyBuilder() {
+      return new ReadOnlyOptionsBuilder(null, null);
+    }
+
+    static ReadWriteOptionsBuilder readWriteBuilder() {
+      return new ReadWriteOptionsBuilder(null, DEFAULT_NUM_ATTEMPTS);
+    }
+  }
+
+  public static final class ReadOnlyOptionsBuilder
+      extends Builder<ReadOnlyOptions, ReadOnlyOptionsBuilder> {
+    @Nullable private TimestampOrBuilder readTime;
+
+    private ReadOnlyOptionsBuilder(@Nullable Executor executor, @Nullable Timestamp readTime) {
+      super(executor);
+      this.readTime = readTime;
+    }
+
+    @Nullable
+    public TimestampOrBuilder getReadTime() {
+      return readTime;
+    }
+
+    public ReadOnlyOptionsBuilder setReadTime(@Nullable TimestampOrBuilder readTime) {
+      this.readTime = readTime;
+      return this;
+    }
+
+    @Nonnull
+    @Override
+    public TransactionOptions build() {
+      final Timestamp timestamp;
+      if (readTime != null && readTime instanceof Timestamp.Builder) {
+        timestamp = ((Timestamp.Builder) readTime).build();
+      } else {
+        timestamp = (Timestamp) readTime;
+      }
+      return new TransactionOptions(
+          executor, EitherReadOnlyOrReadWrite.readOnly(new ReadOnlyOptions(timestamp)));
+    }
+  }
+
+  public static final class ReadWriteOptionsBuilder
+      extends Builder<ReadWriteOptions, ReadWriteOptionsBuilder> {
+    private int numberOfAttempts;
+
+    private ReadWriteOptionsBuilder(@Nullable Executor executor, int numberOfAttempts) {
+      super(executor);
+      this.numberOfAttempts = numberOfAttempts;
+    }
+
+    public int getNumberOfAttempts() {
+      return numberOfAttempts;
+    }
+
+    @Nonnull
+    public ReadWriteOptionsBuilder setNumberOfAttempts(int numberOfAttempts) {
+      Preconditions.checkArgument(numberOfAttempts > 0, "You must allow at least one attempt");
+      this.numberOfAttempts = numberOfAttempts;
+      return this;
+    }
+
+    @Nonnull
+    @Override
+    public TransactionOptions build() {
+      return new TransactionOptions(
+          executor, EitherReadOnlyOrReadWrite.readWrite(new ReadWriteOptions(numberOfAttempts)));
+    }
+  }
+
+  // TODO: Refactor to use Optional when java7 support is dropped
+  static final class EitherReadOnlyOrReadWrite {
```

Comment:
Could this class be merged with TransactionOptions and provide the internal state via package-private methods?

Reply:
Most likely, yes. I was writing this in such a way that it'd be able to take advantage of new java features once we drop support for 7 but it may not be worth it right now.

I'll see what I can do to pull it out.

Reply:
Fixed